runtimevar/awssecretsmanager/awssecretsmanager_test.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/googleapis/gax-go"
+	"gocloud.dev/internal/retry"
+	"gocloud.dev/internal/testing/setup"
+	"gocloud.dev/runtimevar"
+	"gocloud.dev/runtimevar/driver"
+	"gocloud.dev/runtimevar/drivertest"
+)
+
+// This constant records the region used for the last --record.
+// If you want to use --record mode,
+// 1. Update this constant to your AWS region.
+// TODO(issue #300): Use Terraform to get this.
+const region = "us-east-2"
+
+type harness struct {
+	session client.ConfigProvider
+	closer  func()
+}
+
+func newHarness(t *testing.T) (drivertest.Harness, error) {
+	sess, _, done, _ := setup.NewAWSSession(context.Background(), t, region)
+
+	return &harness{
+		session: sess,
+		closer:  done,
+	}, nil
+}
+
+func (h *harness) MakeWatcher(_ context.Context, name string, decoder *runtimevar.Decoder) (driver.Watcher, error) {
+	return newWatcher(h.session, name, decoder, nil), nil
+}
+
+func (h *harness) CreateVariable(ctx context.Context, name string, val []byte) error {
+	svc := secretsmanager.New(h.session)
+	awsName := aws.String(name)
+	token := aws.String("test-awssecretsmanager-client-request-token-create")
+
+	// From AWS Secrets Manager docs:
+	// An asynchronous background process performs the actual secret deletion, so there
+	// can be a short delay before the operation completes. If you write code to
+	// delete and then immediately recreate a secret with the same name, ensure
+	// that your code includes appropriate back off and retry logic.
+	var backoff gax.Backoff
+	if *setup.Record {
+		backoff.Max = 30 * time.Second
+	} else {
+		backoff.Max = time.Millisecond
+	}
+
+	return retry.Call(ctx, backoff,
+		func(err error) bool {
+			if awsErr, ok := err.(awserr.Error); ok {
+				return awsErr.Code() == secretsmanager.ErrCodeInvalidRequestException
+			}
+
+			return false
+		},
+		func() error {
+			_, err := svc.CreateSecretWithContext(ctx, &secretsmanager.CreateSecretInput{
+				Name:               awsName,
+				ClientRequestToken: token,
+				SecretBinary:       val,
+			})
+
+			return err
+		},
+	)
+}
+
+func (h *harness) UpdateVariable(ctx context.Context, name string, val []byte) error {
+	svc := secretsmanager.New(h.session)
+	token := "test-awssecretsmanager-client-request-toke-update-" + string(val)
+
+	_, err := svc.UpdateSecretWithContext(ctx, &secretsmanager.UpdateSecretInput{
+		SecretId: aws.String(name),
+		// ClientRequestToken must be unique per test session but deterministic.
+		ClientRequestToken: aws.String(token),
+		SecretBinary:       val,
+	})
+
+	return err
+}
+
+func (h *harness) DeleteVariable(ctx context.Context, name string) error {
+	svc := secretsmanager.New(h.session)
+
+	_, err := svc.DeleteSecretWithContext(ctx, &secretsmanager.DeleteSecretInput{
+		SecretId:                   aws.String(name),
+		ForceDeleteWithoutRecovery: aws.Bool(true),
+	})
+
```

Comment:
Is there a reasonable RPC to put here so that we could poll until the Delete is complete, instead of retrying in Create? It seems like that would make each tests more hermetic.

Reply:
Yes, the [DescribeSecret](https://aws.amazon.com/premiumsupport/knowledge-center/delete-secrets-manager-secret/) API call could be used to verify that the secret is permanently deleted. Let me do the change.

Reply:
I tried that approach out and it seems I can't make tests reproducible - some tests are failing randomly.

Reply:
This is a bit concerning. Can you dig in a bit more to see why they are not reproducible? The drivertests are enforcing semantic requirements for the driver, and having to make it immutable and using different variable names across tests seem to indicate that something is fishy.

Reply:
I have grabbed the PR and am trying to get the -record tests to pass. It seems like AWS is quite flaky, for example I am able to reproducibly do something like:

* create secret x
* delete secret x
* wait for DescribeSecret to fail
* create secret x
* get secret x
* describe secret x
* get secret x again --> fails!

I'm guessing that these mutating operations are only eventually consistent on the server side, which makes it really hard to write reproducible tests.

Reply:
https://docs.aws.amazon.com/secretsmanager/latest/userguide/troubleshoot_general.html#troubleshoot_general_eventual-consistency

Reply:
OK, I've been able to get the tests to pass now, by modifying each of Create/Update/Delete in the test harness to apply the change, sleep for some time, and then verify the change has been applied (and if not sleep again). I will send a PR to make those changes and add the .replay files after merging this.